### DISCUSSION TITLE
[suggestion] Generic limit to number

### DISCUSSION THREAD

**Discussion by surfsky** [dotnet/csharplang#3496] [Language Ideas] [answered]
``` csharp
bool Add<T>(this T t1, T t2) where  T : number
{
}
```

limit t type to short, int, long, double, float, decimal, nshort, nint, nlong.

**Comment by yaakov-h** [comment, marked as answer]
This is a common case for Shapes (#164) to resolve. AFAIK there's no way to currently implement this without CLR changes or Shape-like witnesses.

**Comment by iam3yal** [comment]
@yaakov-h There are two parts to this so something like Shapes would allow you to have a statically typed T but wouldn't cover the part of forcing a limit over T at compile-time, for example, requiring that the number is within the required range and it also wouldn't allow you the express that in the language because as you well know it's more of the territory of DbC.

**Comment by FiniteReality** [comment]
@yaakov-h Couldn't it be solved in a similar way to Nullable Reference Types' `notnull` constraint? The compiler could simply detect when the generic type parameter is marked with an attribute, say `[Number]` and then issue a warning/error when a numeric type is not passed at the consumption site.

That's however getting very close to "could be a roslyn analyzer" territory, and would probably require extra code to be generated at runtime to enforce the check, since it doesn't stop users from simply disabling the analyzer, or not using it at all.

**Comment by iam3yal** [comment]
@FiniteReality This issue isn't specific to numbers and general/proper solutions exist (#164, #105) so introducing workarounds or temporary solutions into the compiler and runtime may not be desirable.

**Comment by HaloFour** [comment]
@FiniteReality 

Such an approach won't let you treat those values like numbers, though.  For that you need either (or both) runtime or language support.
